### ISSUE TITLE
Windows Event Log Collection Issues

### ISSUE THREAD

**Issue by dokyun-kim** [newrelic/nri-windows-event-logs#5]
Hello,
I tried to collect event log of windowsOS using nri-windows-event-logs, but i have a problem. Can you help me?

Windows event logs are not collected in Insight, and I have rarely modified the configuration file.

Windows OS: Windows Server 2008 R2
NewRelic Infra Agent Version: 1.11.21

- C:\Program Files\New Relic\newrelic-infra\integrations.d\infra-windows-logs-config.yml

```
integration_name: com.newrelic.windows.eventlog

instances:
  - name: eventlog-system
    command: system
  - name: eventlog-application
    command: application
```

- C:\Program Files\New Relic\newrelic-infra\custom-integrations\infra-windows-logs.bat

```
@ECHO OFF
set logName=%1
set eventIds=%2

IF DEFINED eventIds (
    powershell.exe -ExecutionPolicy Bypass -file "C:\Program Files\New Relic\newrelic-infra\custom-integrations\infra-windows-logs.ps1" -LogName %logName% -EventIds %eventIds%
    powershell.exe -ExecutionPolicy Bypass -file "C:\Program Files\New Relic\newrelic-infra\custom-integrations\infra-windows-logs.ps1" -LogName %logName%
)
```

- C:\Program Files\New Relic\newrelic-infra\custom-integrations\infra-windows-logs-definition.yml

```
name: com.newrelic.windows.eventlog
protocol_version: 1
description: On-Host Integration for Windows Event Logs

commands:
  application:
    command:
      - C:\Program Files\New Relic\newrelic-infra\custom-integrations\infra-windows-logs.bat
      - Application
    interval: 30
  system:
    command:
      - C:\Program Files\New Relic\newrelic-infra\custom-integrations\infra-windows-logs.bat
      - System
    interval: 30
```

- C:\Program Files\New Relic\newrelic-infra\custom-integrations\infra-windows-logs.ps1

```
###
# The following command is required for testing:
#
#[System.Diagnostics.EventLog]::CreateEventSource("New Relic","Application")
###


###
# Parameters (a.k.a. Command Line Arguments)
# Usage: -LogName "LogName" -EventIds 4608:4609:4946
# -LogName      The name of the event log to gather events from.  Ex: System, Appication, etc. (Required)
# -EventIds     An optional, colon delimited, list of event ids to gather.
###

param (
    [string]$LogName=$(throw "-LogName is mandatory"),
    [string]$EventIds
)

###
# Logic to handle getting new log entries by saving current date to file
# to use as -After argument of Get-Date in next pull. On first run we use current date.
# On subsequent runs it will use last date written to file.
#
# Uses LogName param to create timestamp for each LogName
###

$LAST_PULL_TIMESTAMP_FILE = "./last-pull-timestamp-$LogName.txt"


###
# If timestamp file exists, use it; otherwise,
# set timestamp to 15 minutes ago to pull some data on
# first run.
###

if (Test-Path $LAST_PULL_TIMESTAMP_FILE -PathType Leaf) {

    $timestamp = Get-Content -Path $LAST_PULL_TIMESTAMP_FILE -Encoding String | Out-String
    $timestamp = [DateTime] $timestamp.ToString()

} else{

    $timestamp = (Get-Date).AddMinutes(-240)

}

###
# Write timestamp to file to pull on next run.
###
Set-Content -Path $LAST_PULL_TIMESTAMP_FILE -Value (Get-Date -Format o)

###
# Pull events using -After param with timestamp
###
$events = Get-EventLog -LogName $LogName -After $timestamp

###
# If event ids were given, filter to keep only the events having an id in our event id list.
###
if ($EventIds) {
    $eventIdStrings = $EventIds.Split(":")
    $eventIdNums = @()
    foreach ($eventId in $eventIdStrings) {
        $eventIdNums += [convert]::ToInt32($eventId)
    }

    # Iterate over the events and copy only those we want to keep into the filteredEvents.
    $filteredEvents = @()
    foreach ($event in $events) {
        if (-Not ($eventIdNums -Contains $event.EventID)) {
            continue
        }
        $filteredEvents += $event
    }
    $events = $filteredEvents
}

###
# Add required 'event_type' to objects from Get-EventLog.
# Add optional 'log_name' value to object.
###


$events.ForEach({
    Add-Member -NotePropertyName 'event_type' -NotePropertyValue 'wineventlog' -InputObject $_
    Add-Member -NotePropertyName 'log_name' -NotePropertyValue $LogName -InputObject $_

});



###
# Create hash table in required format for Infrastructure, populated
# with event object log data and pipe to ConvertTo-Json with
# -Compress argument required in order for Infrastructure to consume.
###
$payload = @{
    name = "com.newrelic.windows.eventlog"
    integration_version = "0.1.0"
    protocol_version = 1
    metrics = @($events)
    inventory = @{}
    events = @()
} | ConvertTo-Json -Compress


###
# Output json string created above with regex to normalize date strings
# post json string conversion. Alternatively, you could create a
# new -NotePropertyName with the proper date string and remove
# the original object property. 
###
Write-Output ($payload -replace '"\\\/Date\((\d+)\)\\\/\"' ,'$1')
```

Thank you

**Comment by a-james-faria** [comment]
@dokyun-kim  Have you checked the infra agent logs, especially on restart of the agent?

**Comment by dokyun-kim** [comment]
Hello,

@a-james-faria 
I restarted the agent and checked the log.
I have seen the error log as below. However, the cause has not been found.

time="2020-05-29T17:13:27Z" level=info msg="Integration health check finished with some errors" instance=eventlog-application integration=com.newrelic.windows.eventlog prefix=integration/com.newrelic.windows.eventlog working-dir="C:\\Program Files\\New Relic\\newrelic-infra\\custom-integrations"


time="2020-05-29T17:13:27Z" level=info msg="Integration health check finished with some errors" instance=eventlog-system integration=com.newrelic.windows.eventlog prefix=integration/com.newrelic.windows.eventlog working-dir="C:\\Program Files\\New Relic\\newrelic-infra\\custom-integrations"


[newrelic-infra.log](https://github.com/newrelic/nri-windows-event-logs/files/4700578/newrelic-infra.log)

Thank you

**Comment by khpeet** [comment]
@dokyun-kim - can you set your newrelic infra agent configuration to briefly collect verbose logs and post them here? Should be `verbose: 1` in the `newrelic-infra.yml` file.

**Comment by dokyun-kim** [comment]
@khpeet 
Hello, Sorry for the late response.

I set it up in the same way and confirmed that it collects event logs normally on windows server 2012 OS.

However, it is still not possible to collect event logs on the Windows Server 2008 R2 OS. I set verbose:1 as you told me and checked the newrelic-infra.log file, and the following error log was printed.

time="2020-07-01T01:09:05Z" level=debug msg="**Missing event_type field for metric**" action=EmitDataSet component=PluginRunner integration=com.newrelic.windows.eventlog metric="map[MachineName:WIN-6QUVT9U5KLE CategoryNumber:0 Message:The New Relic Infrastructure Agent service entered the running state.ReplacementStrings:New Relic Infrastructure Agent running UserName:<nil> Container:<nil> Category:(0) Source:Service Control Manager InstanceId:1.07374886e+09 TimeWritten :1.593565737e+12 Data:110 0 101 0 119 0 114 0 101 0 108 0 105 0 99 0 45 0 105 0 110 0 102 0 114 0 97 0 47 0 52 0 0 0 Index:108304 EntryType:4 EventID:7036 TimeGenerated:1.593565737e+12 Site:<nil>]"

Not supported on Windows Server 2008 R2 OS?

Thank you
